#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import numpy as np
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
data_path = '/content/drive/MyDrive/1. IT study/Data Loves Academy/Data Analyst/10. Модуль 10/data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [ ]:
df

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

Для прибутку в цьому датасеті логічно використовувати `gross income`. Тому що завдання питає саме про прибуток, а `Total` (як можна було б подумати) — це загальна сума чека.

In [6]:
profit_by_branch = df.groupby('Branch')['gross income'].sum().sort_values(ascending=False)
profit_by_branch

,gross income
Branch,
C,5265.1765
A,5057.1605
B,5057.0320


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [8]:
df[df['Branch'] == 'C']['City'].unique()

array(['Naypyitaw'], dtype=object)

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [9]:
invoices_by_category = pd.pivot_table(
    df,
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [10]:
invoices_by_category = pd.pivot_table(
    df,
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

invoices_by_category['Electronic accessories'].idxmax()

print(
    invoices_by_category['Electronic accessories'].idxmax(),
    invoices_by_category['Electronic accessories'].max()
)

A 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


# 1. Чи відрізняється середній чек у членів програми лояльності та звичайних клієнтів?

**Гіпотеза 1**:

**Члени програми лояльності (Member) можуть мати вищий середній чек, ніж звичайні клієнти (Normal).**

Це цікаво для бізнесу, тому що дозволяє оцінити, чи справді membership стимулює клієнтів купувати більше.

In [11]:
df.groupby('Customer type')['Total'].mean().round(2)

,Total
Customer type,
Member,327.79
Normal,318.12


In [12]:
df.groupby('Customer type')['Total'].agg(['mean', 'count', 'min', 'max']).round(2)

,mean,count,min,max
Customer type,,,,
Member,327.79,501,10.68,1042.65
Normal,318.12,499,12.69,1039.29


**Висновок**:

Бачимо, що середній чек Member вищий за Normal, тому програма лояльності потенційно може бути пов'язана з більшою сумою покупки.

**Рекомендація**:

Активніше залучати клієнтів до програми лояльності - через бонуси, персональні знижки, накопичення балів тощо.

# 2. Які категорії товарів приносять найбільший валовий дохід?

**Гіпотеза 2**:

**Не обов'язково категорія, яку купують найчастіше, приносить найбільший прибуток.**

Тут ми дивимося не просто на кількість інвойсів, а на те, які категорії генерують найбільше грошей, тому що кількість покупок і прибутковість - не одне й те саме.

In [13]:
df.groupby('Product line')['gross income'].sum().sort_values(ascending=False).round(2)

,gross income
Product line,
Food and beverages,2673.56
Sports and travel,2624.90
Electronic accessories,2587.50
Fashion accessories,2586.00
Home and lifestyle,2564.85
Health and beauty,2342.56


Для порівняння порахуємо кількість інвойсів по категоріях:

In [14]:
df.groupby('Product line')['Invoice ID'].count().sort_values(ascending=False)

,Invoice ID
Product line,
Fashion accessories,178
Food and beverages,174
Electronic accessories,170
Sports and travel,166
Home and lifestyle,160
Health and beauty,152


In [15]:
df['Product line'].value_counts()

,count
Product line,
Fashion accessories,178
Food and beverages,174
Electronic accessories,170
Sports and travel,166
Home and lifestyle,160
Health and beauty,152


Подивимось, яких товарів продали найбільше за кількістю одиниць:

In [16]:
df.groupby('Product line')['Quantity'].sum().sort_values(ascending=False)

,Quantity
Product line,
Electronic accessories,971
Food and beverages,952
Sports and travel,920
Home and lifestyle,911
Fashion accessories,902
Health and beauty,854


Наприклад:

- Food and beverages — 160 інвойсів, але в цих інвойсах могли купувати по 5–10 товарів;
- Electronic accessories — 170 інвойсів, але в кожному в середньому купували по 2 товари.

Тоді за кількістю інвойсів лідером буде Electronic accessories, а за кількістю проданих одиниць — Food and beverages.

In [17]:
result = df.groupby('Product line').agg({
    'Invoice ID': 'count',
    'Quantity': 'sum',
    'gross income': 'sum'
})

result = result.sort_values('gross income', ascending=False)

result

,Invoice ID,Quantity,gross income
Product line,,,
Food and beverages,174,952,2673.5640
Sports and travel,166,920,2624.8965
Electronic accessories,170,971,2587.5015
Fashion accessories,178,902,2585.9950
Home and lifestyle,160,911,2564.8530
Health and beauty,152,854,2342.5590


Тут `Invoice ID` і `Quantity` означають різні речі.

Для категорії **Food and beverages**:

- **174 Invoice ID** → було 174 покупки/транзакції, у яких була ця категорія.
- **952 Quantity** → загалом у цих покупках було придбано 952 одиниці товарів цієї категорії.

Тобто в одному інвойсі покупець міг придбати не один товар, а, наприклад:

```
Invoice 1 → 3 товари Food and beverages
Invoice 2 → 5 товарів Food and beverages
Invoice 3 → 2 товари Food and beverages
```
Це означає: **174 транзакції**, у яких сумарно купили **952 одиниці товарів**.



**Висновок**

Найбільший сумарний валовий дохід приносить категорія **Food and beverages** - **2673.56**. На другому місці **Sports and travel** - **2624.90**. Найменший валовий дохід має **Health and beauty** - 2342.56.

**Рекомендації**

Катеогорії з найбільшим валовим доходом бізнес потенційно може:

- збільшувати їх асортимент
- забезпечувати достатні запаси
- активніше просувати їх
- розміщувати їх у більш помітних місцях магазину.

Можна підтримувати та розвивати категорію **Food and beverages**, оскільки вона генерує найбільший валовий дохід. Водночас варто дослідити причини низького показника **Health and beauty** - наприклад, асортимент, кількість продажів, середню ціну та кількість товарів у чеку - перш ніж вирішувати, чи потрібно змінювати її асортимент або маркетингове просування.

Варто звернтуи увагу, що це саме сумарний gross income, а не середній прибуток з одного чека. Тому для повнішого висновку варто ще подивитися кількість продажів у кожній категорії.

# 3. Чи відрізняється задоволеність клієнтів залежно від способу оплати?

**Гіпотеза 3**:

**Рейтинг задоволеності клієнтів (Rating) може відрізнятися залежно від способу оплати (Payment).**

Це вже питання не лише що і де купують, а й як купують.

In [18]:
df.groupby('Payment')['Rating'].mean().round(2)

,Rating
Payment,
Cash,6.97
Credit card,7.00
Ewallet,6.95


In [19]:
df.groupby('Payment')['Rating'].agg(['count', 'mean']).round(2)

,count,mean
Payment,,
Cash,344,6.97
Credit card,311,7.00
Ewallet,345,6.95


**Висновок**:

Середній рейтинг клієнтів практично не залежить від способу оплати. Найвищий середній рейтинг мають покупці, які платили кредитною карткою - 7.00, але різниця з готівкою (6.97) та Ewallet (6.95) дуже незначна.

Тому робити висновок, що якийсь спосіб оплати забезпечує кращий клієнтський досвід, на основі цих даних не варто.

**Рекомендація**:

Не варто робити акцент на зміні способів оплати лише для підвищення задоволеності клієнтів, оскільки середні рейтинги за трьома способами оплати майже однакові. Для покращення клієнтського досвіду доцільніше дослідити інші фактори, наприклад категорію товарів, філіал, час покупки або тип клієнта.